### PR TITLE
Move the RemoveProfileNodesAndSpecializeTypes pass to the beginning o…

### PR DESCRIPTION
**Pull request by chengjunlu**
## Motivation
It is found that some optimization passes cannot run successfully with extra prim::profile nodes in the graph in `ProfilingGraphExecutorImpl::runNoGradOptimizations`. (e.g: BatchMM or customPrePasses).
After reviewing the code, it seems the `RemoveProfileNodesAndSpecializeTypes` is a common passes required for both TE and non-TE path.

## Solution
By moving the RemoveProfileNodesAndSpecializeTypes from the TE path to the beginning of `ProfilingGraphExecutorImpl::runNoGradOptimizations` and making it a common passes for all other passes.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

**Comment by ZhiweiYan-96** [comment]
@ZolotukhinM Hi, could you please review our code?

**Comment by ZhiweiYan-96** [comment]
@ZolotukhinM  Hello, do we have any update? : )

**Comment by gujinghui** [comment]
@ezyang  @eellison @ZolotukhinM @malfet

Could you help review and give suggestion. or tell how should review this PR?
Thanks.

**Comment by malfet** [comment]
@gujinghui I'm not familiar with that codepath, but sounds like PR are expected to yield are some perf improvement, but at the same time description lacks any absolute/relative perf numbers for any common models. 

Also, this optimization is currently available for TEfuser.
What were the original motivation to limit it and what has changed?

**Comment by ZhiweiYan-96** [comment]
hi, @malfet, The profiling node may  be inserted between 2 fusable nodes and it blocks the fusion.  Then it will have impact on performance. 
We contemplate that, the `removingprofilingNode` is a common path for TE and non-TE path,  and it can be a common pass.

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -406,6 +406,18 @@ void ProfilingGraphExecutorImpl::runNoGradOptimizations(
...
+    // IR in value types. We're doing such transformation as optimizations
+    // that try to merge/fuse nodes in the graph (e.g. BatchMM and GraphFuser)
+    // work worse in the presence of intermittent prim::profile nodes.
+    // Optimizations relying on the type info are also responsible for
+    // inserting proper type checks. Once we're done with these optimizations
+    // we will wipe the tensor type information from the IR, so that it's not
+    // accidentally used by any other pass.
+    RemoveProfileNodesAndSpecializeTypes(graph);
+    GRAPH_DEBUG(
+        "After RemoveProfileNodesAndSpecializeTypes, before BatchMM\n", *graph);
```
Comment (jgong5) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:419]: nit: it is not "before BatchMM" now.

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**